File: NaryTree/Python/n-ary_tree_preorder_traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> list[int]:
        if not root: return []
        stack, ans = [root], []
        while stack:
            cur = stack.pop()
            if cur :
                stack.extend(reversed(cur.children))
                ans.append(cur.val)
        return ans

File: NaryTree/JS/n_ary_tree_preorder_traversal.js
// Definition for a _Node.
function _Node(val, children) {
   this.val = val;
   this.children = children;
};

/**
 * @param {_Node|null} root
 * @return {number[]}
 */
var preorder = function(root) {
    if (!root) return [];
    const stack = [root], ans = [];
    while (stack.length != 0) {
        const cur = stack.pop();
        if (cur) {
            stack.push(...cur.children.slice().reverse());
            // or
            // const n = cur.children.length;
            // for (let i = n-1; i >= 0; i--) {
            //     stack.push(cur.children[i]);
            // }
            ans.push(cur.val);
        }
    }
    return ans;
};
